### PR TITLE
Snowplow v2

### PR DESCRIPTION
**Pull request by colmsnowplow**
This PR replaces #24261 (Existing changes in there and authorship preserved, refactor added and commit history cleaned up).

Proposed changes for a PR to ampproject/amphtml.

- Added custom context functionality (to be manually entered as SD-JSONs, with a trailing comma)
- Added default amp_id and amp_page_view contexts - sent with every event ([related Iglu Central PR](https://github.com/snowplow/iglu-central/pull/1015)).
- Added custom events.
- Added page pings.
- Added parsing of js-tracker cross-domain linker url value - if found, value is retained for the lifetime of amp-set cookie.
- Added amp linker passing of amp client id to outgoing links.
- Additional standard tracker protocol fields added.
- ampdocurl used for url.
- Domain Userid setting removed (inconsistent with js tracker).

### CODE REVIEW

**Comment by colmsnowplow** [comment]
@googlebot I consent.

**Comment by colmsnowplow** [comment]
Local `gulp pr-check` seems to fail with the below, but doesn't look to be anything to do with my changes.

>     ✗ AssertionError: expected 'i-amphtml-element i-amphtml-layout-responsive i-amphtml-layout-size-defined i-amphtml-layout' to match /i-amphtml-hidden-by-media-query/

Let me know if anything needs fixing!

**Comment by colmsnowplow** [comment]
@max-tgam I think we might need you to comment `@googlebot I consent.` on this one too - apologies for the extra step!

**Comment by max-tgam** [comment]
@googlebot I consent.

**Comment by colmsnowplow** [comment]
@zhouyx Tagging you as the reviewer on the old PR.

**Comment by colmsnowplow** [comment]
Apologies for force-pushing after requesting review - I noticed that I forgot to cherry-pick and squash in a commit. Have given it another once-over and it's definitely ready for review now (and merge if no changes required). :)

File: extensions/amp-analytics/0.1/vendors/snowplow_v2.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    'linker': {
+      'ids': {
+        '_sp_duid': '${duid}',
+        'amp_id': '${ampVistorId}',
+      },
+    },
+  },
+  'cookies': {
+    'enabled': true,
+    'cookieMaxAge': 1800,
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/vendors/snowplow_v2.js]: To confirm, the cookeMaxAge unit is in second. It is the desired behavior to persist cookie for 30 mins?

File: examples/analytics-vendors.amp.html [hunk 2]
```diff
@@ -1687,11 +1777,13 @@
 </div>
 
 <div class="logo"></div>
-<h1 id="top">AMP Analytics</h1>
```
Comment (zhouyx) [inline examples/analytics-vendors.amp.html]: nit: revert

File: examples/analytics-vendors.amp.html [hunk 3]
```diff
@@ -1687,11 +1777,13 @@
...
 <span id="test1" class="box" data-vars-title="Example request with element level overrides">
   Click here to generate an event
 </span>
 <p>
+Example of <a id="ext" data-vars-target-url="../"  data-vars-element-id="home" data-vars-element-content="Internal Link" href="../">Internal Link</a>.
```
Comment (zhouyx) [inline examples/analytics-vendors.amp.html]: Cool, thanks for adding this to test linker feature

File: extensions/amp-analytics/0.1/vendors/snowplow_v2.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/vendors/snowplow_v2.js]: nit: 2019

File: examples/analytics-vendors.amp.html (hunk 2 above)
Comment (colmsnowplow) [reply examples/analytics-vendors.amp.html]: My bad! Fixing.

File: extensions/amp-analytics/0.1/vendors/snowplow_v2.js (hunk 1 above)
Comment (colmsnowplow) [reply extensions/amp-analytics/0.1/vendors/snowplow_v2.js]: Ah no actually I want default value of 1 yr. Thanks for catching, commit on the way to fix.

Comment (zhouyx) [reply extensions/amp-analytics/0.1/vendors/snowplow_v2.js]: Got it, in that case, you don't need to specify the value. The default value is 365 days.

Comment (colmsnowplow) [reply extensions/amp-analytics/0.1/vendors/snowplow_v2.js]: Ah no actually I want default value of 1 yr. Thanks for catching!

**Comment by colmsnowplow** [comment]
Thanks, done - note that I force added the JSON file due to the [gitignore](https://github.com/ampproject/amphtml/blob/0b3fccd4eddcb68403001af7deeaeb57cdf67227/.gitignore#L33).

**Comment by zhouyx** [comment]
>Thanks, done - note that I force added the JSON file due to the gitignore.

Thank you for responding so quickly : ) #26053 had just been merged this morning : )

**Comment by zhouyx** [comment]
@colmsnowplow PR merged. Thank you for the integration!

**Pull request merged**